### PR TITLE
[PEPPER-1429] Refactor ParticipantExitRoute and add tests

### PR DESCRIPTION
**Pull request by denniscunningham**
PEPPER-1429

The fix here was easy (trim input fields), but building appropriate tests was hard, due to Spark Java, which is a poor, outdated framework, and due to the existing structure of the code which was also poor and not conducive to testing.

The main work was in rewriting and testing ParticipantExitRoute -- it was poorly structured, did not handle a handful of cases properly, and mixed route and service code (which makes building tests difficult.) I broke out the exit logic into ParticipantExitService, which enabled testing the route code, and provides a clean break should service tests be added in the future. (I did not add those tests here since that would have taken an additional day, at least, and required additional restructuring downstream. For now we will continue to rely on our PW system test for coverage.)

Since we do not have JUnit tests for ParticipantExitService, I largely kept the code as it was in ParticipantExitRoute, but did some refactoring of `_exitParticipant` since it had a couple obvious deficiencies. I also added some notes where the code intention was unclear.

Other things:

- Renamed `ParticipantNotExist` exception to `EntityNotFound`, and added handling for it in DSMServer, since we did not have a general not found response exception.
- Added a non-static method to UserUtil to allow for mocking. UserUtil is really a service, but at this point there is limited value in setting it up properly and refactoring adjacent code.

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have added regression test labels to the jira ticket
- [ ] I have added verification steps to the jira ticket
- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+public class ParticipantExitService {
+
+    public Map<String, ParticipantExit> getExitedParticipants(String realm) {
+        return ParticipantExit.getExitedParticipants(realm, true);
+    }
+
+    public List<KitDiscard> exitParticipant(String realm, String ddpParticipantId, String userId, boolean inDDP) {
+        _exitParticipant(realm, ddpParticipantId, userId, inDDP);
+
+        // NOTE: The following code was lifted as-is from ParticipantExitRoute.java. I did not attempt to
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java:44]: As noted in the code, this method is pretty much the same as it was in `ParticipantExitRoute`.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        return kitsNeedAction;
+    }
+
+    /**
+     * Exit a participant from DSM and DSS
+     * @param inDDP true if the participant is in DSS. Note: The role of this parameter is unclear. It is passed from
+     *             the front end and the pre-existing code recorded it in the DB, but used
+     *             different criteria to determine if the participant is in DSS. -DC
+     */
+    protected void _exitParticipant(String realm, String ddpParticipantId, String userId, boolean inDDP) {
```
Comment (denniscunningham) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java:70]: This code is refactored a bit so should be compared to the former `exitParticipant` method in `ExitParticipantRoute`.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java [hunk 3]
```diff
@@ -255,7 +255,7 @@ public void onAhStop() {
             }
 
             public void onTerminate() {
-                logger.info("Terminating DSM instance", LogUtil.getAppEngineInstance());
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/DSMServer.java:258]: Thank you

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/UserUtil.java [hunk 4]
```diff
@@ -254,6 +254,10 @@ private static void getList(Connection conn, String query, String role, String u
         }
     }
 
+    public boolean userHasRole(String realm, String userId, String role, String userIdRequest) {
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/UserUtil.java:257]: Why not go full bore and just rename the underlying method?  Happy to put a thumb on that as a separate PR.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/ParticipantExitRoute.java [hunk 5]
```diff
@@ -1,115 +1,75 @@
...
-        if (StringUtils.isNotBlank(realm)) {
-            if (UserUtil.checkUserAccess(realm, userId, ROLE, null)) {
-                return ParticipantExit.getExitedParticipants(realm).values();
-            } else {
-                response.status(500);
-                logger.warn("User with id={} could not access realm={} with role={}", userId, realm, ROLE);
-                return new Result(500, UserErrorMessages.NO_RIGHTS);
-            }
+    public Object processRequest(Request request, Response response, String userId) {
+        String requestMethod = request.requestMethod();
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/ParticipantExitRoute.java:29]: Just to make sure I get it: instead of switching behavior based on the existence of the `realm` param, now we switch based on get/post?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+            } else {
+                //refund label of kits which are not sent yet
+                KitRequestShipping.refundKit(kit.getDsmKitRequestId(), DSMServer.getDDPEasypostApiKey(realm),
+                        maybeInstance.orElse(null));
+            }
+        }
+        return kitsNeedAction;
+    }
+
+    /**
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java:64]: Should this javadoc be on the public method instead of the protected one?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+     * Exit a participant from DSM and DSS
+     * @param inDDP true if the participant is in DSS. Note: The role of this parameter is unclear. It is passed from
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java:66]: `inDDP` looks like it is always set to true in the frontend, as `!ParticipantExitComponent.deletedFromDDP`.  The ability to change this value in the UI has been commented out, and the majority of rows in the table are set to true.

**Review by zyme — APPROVED** [review]
Good stuff!  Nice balance of factoring out code and making it more testable (and tested).

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/ParticipantExitRoute.java (hunk 5 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/ParticipantExitRoute.java:29]: Yes. The former code was relying on an implementation byproduct rather than testing for the request method directly.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java (hunk 6 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java:64]: I mostly wanted to draw attention to `inDDP`, since it's use was odd.
I'll take a crack at the public method, but was hoping to avoid explaining why it returns a list of `KitDiscard` since I do not think it is used in the UI. Although I can hazard a guess that it is kits left dangling by the withdrawal.
Do you know what that's for?

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/UserUtil.java (hunk 4 above)
Comment (denniscunningham) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/UserUtil.java:257]: 'checkUserAccess` is used in many places, so did not want to expand the number of file in the PR.
Although I think `userHasRole` may be more accurate, the method does a bit more checking on input. 
The motivation for changing was to avoid a collision with the existing method name, not to provide a better name, so we would need to resolve that issue to make the change you suggest.

**Review by denniscunningham — COMMENTED** [review]
Thanks for your review @zyme!
A couple questions back to you before I make an update.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java (hunk 6 above)
Comment (zyme) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/participant/ParticipantExitService.java:64]: I don't, but Pegah probably knows.  Either way, l'm up for holding off on this for another time.
